[UI] Add a sign-in button in the app bar

iFlameing opened the pull request:
Related to issue #277 

@helfi92 we only have to remove the profile.displayName from the ListItem in usermenu.jsx and we achieve this feature. Currently ui looks when user is authenticated 
![screenshot from 2019-02-21 21-54-44](https://user-images.githubusercontent.com/33936987/53184924-f8f3e700-3623-11e9-814f-0e9d0ac92b78.png)

and when the user is not authenticated looks like this

![screenshot from 2019-02-21 21-54-55](https://user-images.githubusercontent.com/33936987/53184961-0c9f4d80-3624-11e9-8e8a-666323c50b2f.png)

How I can move this ListItem into seperate file. I have one thought i.e(we can make three separate component and call these component in usermenu.jsx.
1. If the user is not autenticated we can render this component
2. If the user is autenticated but for drawer
3. If the user is authnticated but for Appbar.
)

iFlameing commented:
@helfi92  can you review this now I made the changes as you said :)

iFlameing commented:
@helfi92 you can merge this now.

iFlameing commented:
@helfi92 Do I need to do more modification or it is fine?

helfi92 commented:
> @helfi92 Do I need to do more modification or it is fine?

Sorry, I typically don't do work during the weekend. I haven't looked at it yet. I will look at it now :)

iFlameing commented:
@helfi92 I fixed both of them you can merge this now.

iFlameing commented:
@helfi92 I tried my best to refactor the code. can you review this or make some comment.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,112 @@
+    this.props.client.clearStore();
+  };
+
+  MenuClick = e => {
```
Rename to `handleMenuClick`.

iFlameing replied:
when I am renaming to handleMenuClick eslint give me an error 
` prop key to for handleMenuClick must begin with on`
what should I do?

iFlameing replied:
sorry I got it.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,112 @@
+    anchorEl: null,
+    signInDialogOpen: false,
+  };
+
```
Define a `static propTypes` and `static defaultProps` methods. See the other component files for inspiration.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,112 @@
+            signInDialogOpen={signInDialogOpen}
+            SignInDialogClosehandler={this.SignInDialogClose}
+            SignInDialogOpenhandler={this.SignInDialogOpen}
+            menuClickhandler={this.MenuClick}
```
Rename to `onMenuClick={this.handleMenuClick}`.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,112 @@
+            signInDialogOpen={signInDialogOpen}
+            SignInDialogClosehandler={this.SignInDialogClose}
+            SignInDialogOpenhandler={this.SignInDialogOpen}
+            menuClickhandler={this.MenuClick}
```
See comment above.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,112 @@
+          </MenuItem>
+          <MenuItem
+            title={`Sign Out of ${process.env.APPLICATION_NAME}`}
+            onClick={this.handleClickSignOut}>
```
Rename to `handleSignoutClick`.

iFlameing commented:
@helfi92 I made all the changes which are required.

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,84 @@
+  icon: {
+    fill: theme.palette.common.white,
+    marginRight: theme.spacing.unit,
+    marginLeft: theme.spacing.unit,
```
Remove both `marginLeft` and `marginRight`. Add a marginLeft on the `UserMenuButton` instead. See other comment.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,123 @@
+    overflow: 'hidden',
+    whiteSpace: 'nowrap',
+    textOverflow: 'ellipsis',
+  },
```
Add a new class name:

```jsx
userMenuButton: {
  marginLeft: theme.spacing.unit,
},
```

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,123 @@
+          />
+        ) : (
+          <UserMenuButton
+            user={user}
```
Add `className={classes.userMenuButton}`. In order for the class to be applied, you will need to spread the props inside the `UserMenuButton` file. See other comment.

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,84 @@
+      signInDialogOpen,
+      onSignInDialogOpen,
+      onSignInDialogClose,
+      onMenuClick,
```
Spread the rest of the props so we can give it to the button component:

```jsx
// ...
onMenuClick,
...props
```

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,84 @@
+            variant="contained"
+            color="primary"
+            className={classes.icon}
+            onClick={onSignInDialogOpen}>
```
Add `{...props}` at the end so the class name can be applied.

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,84 @@
+          <Button
+            variant="contained"
+            color="primary"
+            className={classes.icon}
```
I don't think we need `classes.icon`. Remove this line.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,123 @@
+import UserMenuButton from './UserMenuButton';
+
+@withStyles(theme => ({
+  avatar: {
```
This is not being used. Remove unused classes from the rest.

iFlameing commented:
@helfi92  is this possible to do because we want the same time to render userMenuList and userMenuButton from the userMenu component when docs is true.  can you tell me how I can proceed towards it?

helfi92 commented:
I'm not sure I understand the question but basically the docs site should just have the same behavior as the normal site, that is to use

![screen shot 2019-02-27 at 11 00 13 am](https://user-images.githubusercontent.com/3766511/53503849-e3951780-3a7e-11e9-9cfc-3c1aee66f4f8.png)

for the left drawer. Let me know if that doesn't answer your question.

iFlameing commented:
@helfi92 Mainly we have two components for showing signin button one UserMenuList(which is render when the left drawer is open) and other userMenuButton(which is rendered when a drawer is closed). 

The problem is in the documentation view where the drawer is permanent.

 In the permanent drawer, we want to render userMenuList component(in left side of taskcluster).

 And in appBar we want to render userMenuButton(on the left top) at the same time.

 But we can only render one component either userMenuButton or userMenuList at one time.

iFlameing commented:
@helfi92 I resolve the problem I will make a commit tommorrow.

helfi92 commented:
> one UserMenuList(which is render when the left drawer is open)

Correct

> and other userMenuButton(which is rendered when a drawer is closed).

The userMenuButton has nothing to do with the left drawer. It should always be located top right in the app bar. If you open the the drawer, then we should be able to see both the UserMenuList in the drawer and UserMenuButton in the app bar.

> And in appBar we want to render userMenuButton(on the left top) at the same time

.Top right, not top left.

> But we can only render one component either userMenuButton or userMenuList at one time

Why is that?

iFlameing commented:
@helfi92 I push a new commit with all the changes as you suggested above. And please make comment if something has to be modified :)

iFlameing commented:
@helfi92 I fix the above issue.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,109 @@
+export default class UserMenu extends Component {
+  static defaultProps = {
+    user: '',
+    navOpen: null,
```
Remove `navOpen`.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,109 @@
+  static defaultProps = {
+    user: '',
+    navOpen: null,
+    appBar: null,
```
`appBar: false,`

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,109 @@
+
+  static propTypes = {
+    user: string,
+    navOpen: bool,
```
Remove `navOpen`.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,109 @@
+  };
+
+  render() {
+    const { classes, user, navOpen, appBar } = this.props;
```
Remove `navOpen`.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,109 @@
+
+    return (
+      <Fragment>
+        {navOpen && appBar ? (
```
Remove `navOpen`.

iFlameing commented:
@helfi92 I modify the code as you suggested above. can you merge this now :).

thanks for all the help in creating pr.

helfi92 commented:
Looking at this now :)

iFlameing commented:
@helfi92 I made all the above changes please take look!

iFlameing commented:
@helfi92 I pushed the commit with the above modification :)

iFlameing commented:
@helfi92 when I am rebasing it automatically resolve some code and change the code to previous commit . How I resolve this?

helfi92 commented:
When the rebase stop in the middle, you will need to resolve the change. Once that is resolved, you can continue with:

```bash
git add <file-changed>
git rebase --continue
```

iFlameing commented:
@helfi92 I rebased all the pr :) sorry for package.lock.json I already updated my master earlier with serialize-error module which is missing when I am running web-server in master branch so I install it and for changing branch, I commited in the master. I don't why this happen. can you merge this now :)

helfi92 commented on services/web-server/package-lock.json:
```diff
@@ -0,0 +1,31 @@
+{
```
Remove this file.

helfi92 replied:
```
rm <path/to/file>
git rm <path/to/file>
git commit -m 'Remove package-lock.json'
```

iFlameing commented:
@helfi92 How I can remove this. I don't know?

helfi92 commented:
https://github.com/taskcluster/taskcluster/pull/280#discussion_r261745582.

iFlameing commented:
sorry for disturbing you so much. thanks for your help. I learned a lot from this conversation.

iFlameing commented:
@helfi92 now you can merge this all test has been passed and all the conversation is resolved. thanks!

helfi92 commented on ui/src/components/Dashboard/index.jsx:
```diff
@@ -286,22 +286,18 @@ export default class Dashboard extends Component {
                 </IconButton>
               </Tooltip>
             )}
+            <UserMenu navOpen appBar />
```
Interesting. I thought we had removed `navOpen`?

iFlameing replied:
This problem occur when I have done rebase. It automatically got resolve adding this.

iFlameing replied:
No, navOpen is being used in the code. see the below comment I put the reference of the code.

iFlameing replied:
Going to investigate and if it is not needed I will remove it.

helfi92 replied:
Remove `navOpen`.

helfi92 commented on ui/src/components/Dashboard/index.jsx:
```diff
@@ -243,10 +243,10 @@ export default class Dashboard extends Component {
         </div>
         <Divider />
-        <UserMenu />
+        {docs ? <UserMenu /> : <UserMenu navOpen />}
```
Is `navOpen` being used?

iFlameing replied:
yes it is being used here you can refer previous commit  https://github.com/taskcluster/taskcluster/pull/280/commits/2f8194b1e4eb994e9bd613096cb77d997b682414

helfi92 replied:
I still don't see how it's being used. What is it doing? Didn't we remove it in https://github.com/taskcluster/taskcluster/pull/280#discussion_r261261456?

helfi92 replied:
Replace this line with `<UserMenu />`.

iFlameing commented:
@helfi92 I think it is still not solved because after adding your changes the right avatar not position correctly in the center. please check it out.
![screenshot from 2019-03-02 04-49-03](https://user-images.githubusercontent.com/33936987/53671954-95faf380-3ca6-11e9-8268-47af111443ca.png)

helfi92 commented:
I have to sign off now but I'll be back in a week to hopefully merge this :) I just don't want to merge something right now and not be available if anything goes wrong while I'm away. Sorry for that. Awesome work!

iFlameing commented:
@helfi92 I think rebase done some terrible work it resolve some file automatically which lead to this problem. I really tried to make this merge today but I think I can close this pr and make a new pr with all the change you suggested me in this pr. so that it is easy to review and then we will merge it.

helfi92 commented:
Works with me :)

iFlameing commented:
@helfi92 It works in firefox but not in chrome.

iFlameing commented:
@helfi92 It is the problem related to padding in chrome.

djmitche commented:
@iFlameing @helfi92 is away this week.  I'm less familiar with this code, but can do a review when you believe it is ready to merge.  As for differences between browsers -- I bet you know more than i do about that, so I will leave you to puzzle it out.  Let me know when you'd like a final review.

iFlameing commented:
@djmitche I think it is ready for final review. I will open a new pr for the browser issue when this is merge.
thanks!

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,92 @@
+  },
+  avatarButton: {
+    height: 6 * theme.spacing.unit,
+    width: 6 * theme.spacing.unit,
```
Add `padding: 0` to center the button.

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -0,0 +1,92 @@
+
+@withStyles(theme => ({
+  icon: {
+    fill: theme.palette.common.white,
```
Is this having any effect? I think we can remove it.

iFlameing replied:
yes, we can remove this.

iFlameing commented:
@helfi92 I made the changes as you said and also the padding issue is solved in chrome browser :)

helfi92 commented on ui/src/components/UserMenu/UserMenuButton.jsx:
```diff
@@ -12,12 +12,10 @@ import SignInDialog from '../SignInDialog';
   avatarButton: {
     height: 6 * theme.spacing.unit,
     width: 6 * theme.spacing.unit,
+    padding: '0',
```
Remove quotations. `padding: 0`.

iFlameing replied:
I removed the quotations.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,129 @@
+  },
+}))
+@withAuth
+@withApollo
```
Let's make `@withAuth` and `@withApollo` come before `@withStyles`.

helfi92 commented on ui/src/components/UserMenu/index.jsx:
```diff
@@ -0,0 +1,129 @@
+            buttonProps={{
+              classes: { containedPrimary: classes.buttonContainedPrimary },
+            }}
+            className={classes.userMenuButton}
```
```jsx
import classNames from 'classnames';

// ...

className={classNames({ [classes.userMenuButton]: !user })}
```

iFlameing commented:
@helfi92 can you merge this. I pushed the modify changes :)

iFlameing commented:
@helfi92 Adding these changes remove this problem. please! take a look.

--- a/ui/src/components/UserMenu/UserMenuButton.jsx
+++ b/ui/src/components/UserMenu/UserMenuButton.jsx
@@ -19,6 +19,9 @@ import { withAuth } from '../../utils/Auth';
     width: 6 * theme.spacing.unit,
     padding: 0,
   },
+  Button: {
+    whiteSpace: 'nowrap',
+  },
 }))
 export default class UserMenuButton extends Component {
   static propTypes = {
@@ -49,7 +52,7 @@ export default class UserMenuButton extends Component {
       return (
         <Fragment>
           <Button
-            className={className}
+            className={classNames({ [classes.Button]: !user })}
             variant="contained"
             color="primary"
             onClick={onSignInDialogOpen}

iFlameing commented:
@helfi92 I like the changes and pushed the modify code :)

helfi92 commented:
Awesome, I will take a look soon!
